Feature/#15 review

chanwooo opened the pull request:
review 서비스 구현
테스트 일부 작성중

reasonSun committed:
test/#8_대기열 테스트 작성

chanwooo committed:
test/#8_MakeDummy
dummy 만드는 클래스 추가

chanwooo committed:
refactor/#12_도메인별 패키지 추가, 서비스 레이어 생성

chanwooo committed:
refactor/#12_임시저장

chanwooo committed:
refactor/#12_waitingQueue -> waiting으로 이름 변경

chanwooo committed:
refactor/#12_큐잉서비스 request, response 작성

chanwooo committed:
refactor/#12_review, rating - request, response 작성

chanwooo committed:
minor/type변경, 변수이름수정

reasonSun committed:
refactor/#12_menu,store,user - request,response 작성

reasonSun committed:
Merge remote-tracking branch 'origin/feature/#12_서비스별_패키지_나누기' into feature/#12_서비스별_패키지_나누기

chanwooo committed:
review반영/self-descriptive한 method명 updateRating->updateAll()

chanwooo committed:
review반영/msg -> message

chanwooo committed:
review반영/
- Float, Integer -> primitive type으로 변경
- Lombok Annotation 적용
- Review.java : 변수 이름 변경 member -> user

chanwooo committed:
review반영/구현안된기능에 UnsupportedOperationException 적용

chanwooo committed:
review반영/Lombok Annotation 적용, default값 제거

reasonSun committed:
review반영/
- Lombok Annotation 적용
- image -> imageUrl

reasonSun committed:
review반영/Entity Lombok Annotation 적용

chanwooo committed:
feature/리뷰반영
- import * 제거
- Entity에 NoArgsConstructor access level protected 적용
- wrapped type 제거, @NonNull 제거

chanwooo committed:
Merge remote-tracking branch 'origin/feature/#12_서비스별_패키지_나누기' into feature/#12_서비스별_패키지_나누기

chanwooo committed:
feature/리뷰반영
- fix member -> user
- fix Integer -> int

chanwooo committed:
feature/soft delete를 위한 필드 추가

chanwooo committed:
feature/필드 이름 변경image->imageUrl

chanwooo committed:
feature/reviewService 구현

chanwooo committed:
#15 - review 현재 사용하지 않는 테스트 파일 삭제

chanwooo committed:
#15 - review test용 DummyDataMaker

chanwooo committed:
#15 - review store대신 waiting을 멤버로 변경
hide(soft delete) 구현
update 구현(이전 기록을 hide, 새로운 review 추가)
service에 find 추가

chanwooo committed:
#15 - review service - create test 작성

chanwooo committed:
#15 - review - 사용하지않는 파일 삭제

chanwooo committed:
#15 - review - WaitingRepository.java 추가

chanwooo committed:
#18-Review Store->Waiting 변경으로 find메서드 이름 변경

chanwooo committed:
#18-Review BaseEntity 패키지 위치 변경 user->common
BaseEntity 변수명 변경 Id -> id

chanwooo committed:
#18-Review User table name 지정 users(h2 예약어)

f-lab-trey commented on src/main/java/flab/quing/review/Rating.java:
```diff
@@ -0,0 +1,28 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@Entity
+public class Rating extends BaseEntity {
```
[질문]
Store의 평균 별점 개념의 Entity인거 같은데, Store와 `ManyToOne` 관계인 것이 잘 이해가 안됩니다. 조금 더 부연설명 부탁드립니다.

chanwooo replied:
연관관계가 있는 필드라고 생각하고 아무생각없이 `ManyToOne`을 달아버린거같아요 
1:1이 되어야할거같아요 `OneToOne`으로 수정하겠습니다

f-lab-trey commented on src/main/java/flab/quing/review/Rating.java:
```diff
@@ -0,0 +1,28 @@
+
+import javax.persistence.Entity;
+import javax.persistence.ManyToOne;
+
+@Data
+@Setter(AccessLevel.NONE)
+@Builder
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@Entity
```
이전에도 드렸던 피드백인데, 이만큼이 모든 Entity에 중복으로 달려있으니까 Custom annotation으로 만들어서 중복을 없애보면 좋을 거 같습니다.

chanwooo replied:
`@DataEntity` 이런식의 어노테이션을 만들고 
그안에
```java
@Data
@Setter(AccessLevel.NONE)
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Entity
```
이것들을 넣어주는식으로하고 싶었는데 생각대로 잘 안되네요. 뭔가 원하는 예제도 잘 검색이 안되는데.. 좀 더 찾아봐야할거같아요.

f-lab-trey replied:
그렇군요.. 저도 묶음 annotation을 만들어본적은 없어서 되는지 잘 모르겠어요. 자료가 안나오면 그냥 넘어가도 괜찮습니다.

f-lab-trey replied:
merge annotation으로 검색하니까 조금 나오네요. e.g. `@SpringBootApplication `

f-lab-trey commented on src/main/java/flab/quing/review/Review.java:
```diff
@@ -0,0 +1,60 @@
+
+    private String message;
+
+    private String imageUrl;
```
Review에 사진이 꼭 하나만 들어갈까요?

chanwooo replied:
굳이 한개로 제한할 이유가 없을거같아요 .
사진을 추가 수정 삭제 하기 위한 코드도 추가해야겠네요

f-lab-trey commented on src/main/java/flab/quing/review/Review.java:
```diff
@@ -0,0 +1,60 @@
+    private String imageUrl;
+
+    public void of(ReviewRequest reviewRequest) {
+
```
[minor]
불필요한 newline인거 같습니다.

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+
+        User user = waiting.getUser();
+
+        Review review = Review.builder()
+                .user(user)
+                .waiting(waiting)
+                .build();
+        review.of(reviewRequest);
```
`Review` object를 두 단계로 나눠서 만드는 것보다 한 번에 만드는 것이 좋아보입니다. `Review.of`에서 `User`를 paramter로 받으면 가능할 거 같습니다.

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public ReviewResponse update(ReviewRequest reviewRequest) {
+        Waiting waiting = waitingRepository.findById(reviewRequest.getWaitingId()).get();
+        Review review = reviewRepository.findByWaitingOrderByIdDesc(waiting).get();
```
`reviewRepository.findByWaitingOrderByIdDesc`이 `Optional`를 리턴할텐데, Empty이면 `.get()` 할 때 예외가 발생할 거 같네요. 예전에 테스트 연습한 코드에서 `Optional` handling 하는 코드를 참고해보세요.

chanwooo replied:
코드를 작성할때도 이부분에 대해 고민을 좀 했어요.
비어있다면 어떻게 하는게 좋은가?

비어있는 객체를 만들어서 채워줘야할까?
`orElseGet(Waiting::new)` -?

NoSuchElement 같은 익셉션을 띄워야할까?
`orElseThrow(()->new NoSuchElementException())` -?
를 해야하는건가 생각을 했었고

get()을 사용하는것만으로도 가져온게 없다면 `NoSuchElementException` 을 던질거라
이걸로 충분하다 생각해서 get을 그대로 사용했어요.

`.orElseThrow` 를 이용해서 어떤 exception을 던질지를 좀 더 명시적으로 하는게 좋을까요?

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+    public ReviewResponse update(ReviewRequest reviewRequest) {
+        Waiting waiting = waitingRepository.findById(reviewRequest.getWaitingId()).get();
+        Review review = reviewRepository.findByWaitingOrderByIdDesc(waiting).get();
+        hide(review.getId());
```
이 줄은 무엇을 하는건가요?

chanwooo replied:
update를 구현할때 기존 기록을 남기고 하고싶어서
가장 최근 review를 가져와서 숨기기(hide)를 하고
새로 create를 하는방식으로 구현하려고 했어요.

review_id, message
3, ㅋㅋㅋ -> hide
4, ㅎㅎㅎ -> create

f-lab-trey replied:
[opinion]
변경 History를 남기는 것은 Audit 관련 기능을 이용해 따로 처리할 수 있어서 business logic에 이 부분이 따로 들어가는 것은 적절치 않아 보입니다.

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public ReviewResponse get(long reviewId) {
+        return reviewRepository.findById(reviewId).get().toResponse();
```
Ditto regarding `Optional`

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public ReviewResponse find(long waitingId) {
+        Waiting waiting = waitingRepository.findById(waitingId).get();
```
ditto. 아래에도 `Optional` handling이 안되어 있는 것으로 보입니다. 중복이니까 여기까지만 코멘트하겠습니다. 모두 찾아서 수정해주세요!

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public List<ReviewResponse> getList(long storeId) {
+        Store store = storeRepository.findById(storeId).get();
+        return reviewRepository.findAllByWaitingStoreAndDeletedIsFalse(store).stream().map(Review::toResponse).collect(Collectors.toList());
```
[minor]
`stream` chain은 한 줄당 하나의 operation만 쓰는 것이 더 보기 좋습니다.

f-lab-trey commented on src/main/java/flab/quing/review/dto/RatingResponse.java:
```diff
@@ -0,0 +1,20 @@
+
+    private float rating;
+
+}
```
[minor]
아래에 newline 넣어주세요.

f-lab-trey commented on src/main/java/flab/quing/review/dto/ReviewRequest.java:
```diff
@@ -0,0 +1,23 @@
+
+    private String message;
+
+    private String imageUrl;
```
이미지가 여러 개라면?

f-lab-trey commented on src/main/java/flab/quing/user/UserService.java:
```diff
@@ -0,0 +1,18 @@
+
+    StoreManagerResponse storeSignUp(StoreMangerRequest storeMangerRequest);
+
+    StoreManagerResponse storeSignIn(String name, String password);
```
이 password는 암호화 된 문자열이겠죠? 만약 그렇다면, 이름을 `encryptedPassword`로 바꾸는 것이 어떨까요?

f-lab-trey commented on src/test/java/flab/quing/review/ReviewServiceImplTest.java:
```diff
@@ -0,0 +1,121 @@
+
+
+        //when
+        when(waitingRepository.findById(anyLong())).thenReturn(Optional.of(waiting));
+
+        when(reviewRepository.save(any(Review.class))).thenReturn(review);
```
이 `when`은 stubbing 이니까 given 블록에 들어가는 것이 더 적절하다고 생각합니다.

chanwooo committed:
#15-Review blank line 정리

chanwooo committed:
#15-Review of method static으로 변경, user, waiting 파라미터 추가

chanwooo committed:
#15-Review Store-Ratring 관계 OneToOne으로 변경

chanwooo committed:
#15-Review 줄바꿈 변경

chanwooo commented:
바로 할수있는 부분들 위주로 우선 리뷰 반영하였습니다!

- Store ManyToOne -> OneToOne
- @DataEntity (작업중)
- Review imageUrl -> List<String> imageUrls (작업중)
  - 관련 클래스들에서도 list를 추가 수정 삭제 할수 있도록 변경
- 줄바꿈 관련 정리
- of 메서드에 파라미터로 user, waiting 추가, static으로 변경
- Optional.get() (고민중)
- password -> encryptedPassword 유선님한테 요청
- when() //given 쪽으로 이동

chanwooo committed:
#15-Review optional.get() 처리
Exception 생성
- NoSuchReviewException
- NoSuchStoreException
- NoSuchWaitingException

chanwooo committed:
#15-Review imageUrl -> List로 변경

chanwooo committed:
#15-Review waiting으로 find 해올때 1개의 결과만 가져오도록 limit 설정

chanwooo committed:
#15-Review 기본생성자 제거

chanwooo committed:
#15-Review repository, service test 작성

f-lab-trey commented on src/main/java/flab/quing/common/BaseEntity.java:
```diff
@@ -9,14 +9,20 @@
 @Getter
+@Setter
+@MappedSuperclass
 public class BaseEntity {
```
[minor]
`BaseEntity`가 Entity package 안에 있는 것도 괜찮아 보입니다. 보통 common 같은 package에는 util 클래스 같은 것들이 들어갑니다.

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+    @Transactional
+    @Override
+    public ReviewResponse hide(long reviewId) {
+        Review review = reviewRepository.findById(reviewId).orElseThrow(() -> new NoSuchReviewException());
```
method chain은 가능하면 한 줄에 하나씩 쓰는 것이 읽기 쉽습니다.

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Override
+    public ReviewResponse get(long reviewId) {
```
method 이름을 `getByReviewId`로 바꾸는 것이 어떨까요?

f-lab-trey commented on src/main/java/flab/quing/review/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Override
+    public ReviewResponse find(long waitingId) {
```
method 이름을 `getByWatingId` 또는 `findByWaitingId`로 바꾸는 것이 어떨까요?

f-lab-trey commented on src/main/java/flab/quing/waiting/QuingService.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.List;
+
+//Chanwoo
```
이제는 불필요한 comment 같습니다.

f-lab-trey commented on src/main/java/flab/quing/waiting/QuingService.java:
```diff
@@ -0,0 +1,25 @@
+
+    WaitingResponse append(WaitingRequest waitingRequest);
+
+    List<WaitingResponse> getList(Long storeId);
```
[질문]
Boxed type을 사용하는 것은 `null`이 parameter로 넘어올 수도 있단는건가요?

f-lab-trey commented on src/main/java/flab/quing/waiting/WaitingQueueRepository.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.List;
+
+public interface WaitingQueueRepository extends JpaRepository<Waiting, Long> {
+    List<Waiting> findAllByStore(Store store);
```
Interface 사용자 관점에서, WaitingQueue를 얻기 위해 `store` object 전체를 넘기는 것보다는 `storeId`를 넘기는 것이 더 편하지 않을까요?

chanwooo committed:
#15 BaseEntity 패키지 변경 common -> entity

chanwooo committed:
#15 불필요한 주석 제거

chanwooo committed:
#15 QuingService 파라미터 타입 변경 Long->long

chanwooo committed:
#15 repository에서 객체의 id로 조회하게 변경

chanwooo committed:
#15 ReviewService 메소드이름 변경
- get -> getByReviewId
- find -> getByWaitingId

chanwooo committed:
#15 ReviewServiceImpl update 구현 변경, repository에서 객체id로 조회하도록 변경

chanwooo commented:
리뷰 반영했습니다.
- baseEntity 패키지 변경 common -> entity
- update 구현 변경
- method chain 줄바꿈
- reviewService get, find -> getByReviewId, getByWaitingId
- storeService 파라미터 타입 Long -> long
- waitingRepository storeId로 조회하도록 변경

waiting에 관련된 부분은 이 브랜치가 머지된 이후에 작업하려해요.

f-lab-trey commented:
commit을 잘 구분해주셔서 리뷰하기 편했습니다. 👏🏼

chanwooo committed:
컨플릭트 해결을 위해 develop을 머지, 수정

chanwooo committed:
컨플릭트 해결을 위해 develop을 머지, 수정

chanwooo committed:
컨플릭트 해결을 위해 develop을 머지, 수정

chanwooo committed:
컨플릭트 해결을 위해 develop을 머지, 수정
